File: techniques/rates-proportions/r/rates_proportions.R
# Summary statistics for rates and proportions (Reference §1.8)
# From-scratch base-R plus idiomatic calls (binom.test, poisson.test, prop.test).
# Run with:  Rscript rates_proportions.R
#
# Inputs used below:
#   cases, population        : numerator and denominator for prevalence
#   new_cases, at_risk_start : for incidence proportion (risk)
#   follow_up_times          : numeric vector of per-subject follow-up durations
#   events, person_time_total: event count and total at-risk time, for incidence rate
#   x, n                     : successes / trials for proportion CIs
#   conf                     : confidence level (e.g. 0.95 -> 95% CI)

prevalence            <- function(cases, population) cases / population
incidence_proportion  <- function(new_cases, at_risk_start) new_cases / at_risk_start
person_time           <- function(follow_up_times) sum(follow_up_times)
incidence_rate        <- function(events, person_time_total) events / person_time_total

# --- CIs for a proportion p = x / n -------------------------------------
ci_wald <- function(x, n, conf = 0.95) {
  p <- x / n; z <- qnorm(0.5 + conf / 2); h <- z * sqrt(p * (1 - p) / n)
  c(max(0, p - h), min(1, p + h))
}
ci_wilson <- function(x, n, conf = 0.95) {
  p <- x / n; z <- qnorm(0.5 + conf / 2); z2 <- z^2
  denom <- 1 + z2 / n
  center <- (p + z2 / (2 * n)) / denom
  half <- (z / denom) * sqrt(p * (1 - p) / n + z2 / (4 * n^2))
  c(max(0, center - half), min(1, center + half))
}
ci_clopper_pearson <- function(x, n, conf = 0.95) {
  a <- 1 - conf
  lo <- if (x == 0) 0 else qbeta(a / 2, x, n - x + 1)
  hi <- if (x == n) 1 else qbeta(1 - a / 2, x + 1, n - x)
  c(lo, hi)
}

# --- exact Poisson CI for a rate ----------------------------------------
ci_poisson_rate <- function(events, person_time_total, conf = 0.95) {
  a <- 1 - conf
  lo_count <- if (events == 0) 0 else qchisq(a / 2, 2 * events) / 2
  hi_count <- qchisq(1 - a / 2, 2 * events + 2) / 2
  c(lo_count, hi_count) / person_time_total
}

# Library: stats::binom.test (Clopper-Pearson), stats::prop.test (Wilson),
#          stats::poisson.test (exact Poisson rate CI); epitools::pois.exact, binom::binom.confint
library_demo <- function(x, n, events, pt) {
  cat("binom.test (Clopper-Pearson):", binom.test(x, n)$conf.int, "\n")
  cat("prop.test (Wilson, correct=FALSE):", prop.test(x, n, correct = FALSE)$conf.int, "\n")
  cat("poisson.test rate CI:", poisson.test(events, pt)$conf.int, "\n")
}

if (sys.nframe() == 0) {
  cat("=== Prevalence ===\n120 / 5000 =", prevalence(120, 5000), "\n")
  follow <- c(2, 5, 1.5, 5, 3.2, 5, 0.8, 4.1); pt <- person_time(follow)
  cat("\n=== Incidence ===\nperson-time:", pt, "  rate (3 events):", incidence_rate(3, pt),
      "  incidence proportion (3/8):", incidence_proportion(3, 8), "\n")
  cat("\n=== 95% CIs for x=8, n=100 ===\n")
  cat("Wald           :", ci_wald(8, 100), "\n")
  cat("Wilson         :", ci_wilson(8, 100), "\n")
  cat("Clopper-Pearson:", ci_clopper_pearson(8, 100), "\n")
  cat("\n=== 95% exact Poisson CI: 3 events /", pt, "person-years ===\n")
  cat("rate =", 3 / pt, "  CI =", ci_poisson_rate(3, pt), "\n")
  cat("\n--- library ---\n"); library_demo(8, 100, 3, pt)
}

File: techniques/rates-proportions/python/rates_proportions.py
"""Summary statistics for rates and proportions (Reference §1.8).

Measures of how often an event occurs in a population:

  - prevalence            : proportion WITH a condition at a point in time (cross-sectional)
  - incidence proportion  : "risk" = proportion DEVELOPING the condition over a period
  - incidence rate        : events per person-time at risk (handles variable follow-up)
  - person-time           : sum of individual at-risk follow-up durations

Confidence intervals:
  for a proportion p = x/n
    - Wald (normal approx)    : p +/- z * sqrt(p(1-p)/n)        (poor near 0/1 or small n)
    - Wilson score            : recommended default
    - Clopper-Pearson (exact) : guaranteed >= nominal coverage; uses the Beta distribution
  for a rate lambda = events / person_time
    - exact Poisson CI        : via the chi-squared / Gamma relationship
"""
from __future__ import annotations    # stdlib: postpone type-hint evaluation (lets us write int | None)

import math    # stdlib: scalar math (sqrt, log, exp, comb, lgamma, pi, ...)
from typing import NamedTuple, Sequence    # stdlib: type hints (Sequence = indexable iterable; NamedTuple = tuple with named fields)

from scipy import stats    # distributions, hypothesis tests, PPFs (norm, t, chi2, ttest_ind, ...)


class CI(NamedTuple):
    """Confidence interval. Unpacks like a tuple: ``lo, hi = ci_wilson(x, n)``."""
    lower: float
    upper: float


# --------------------------------------------------------------------------
# Point estimates
# --------------------------------------------------------------------------
def prevalence(cases: int, population: int) -> float:
    """Point prevalence = cases / population (cross-sectional)."""
    return cases / population


def incidence_proportion(new_cases: int, at_risk_at_start: int) -> float:
    """Cumulative incidence ("risk") = new_cases / at_risk_at_start over the follow-up period."""
    return new_cases / at_risk_at_start


def person_time(follow_up_times: Sequence[float]) -> float:
    """Total person-time = sum of individual at-risk follow-up durations.

    ``follow_up_times`` is a sequence with one entry per subject (e.g. years observed).
    """
    return float(sum(follow_up_times))


def incidence_rate(events: int, person_time_total: float) -> float:
    """Incidence rate = events / total person-time (e.g. per person-year)."""
    return events / person_time_total


# --------------------------------------------------------------------------
# Confidence intervals for a proportion
# --------------------------------------------------------------------------
# Conventions for the CI functions below:
#   x    -- number of successes (count of the event of interest)
#   n    -- sample size (total trials)
#   conf -- confidence level (e.g. 0.95 for a 95% CI)
# All return a CI named tuple with fields ``lower`` and ``upper``, clipped to
# [0, 1]. Unpacks like a tuple, so ``lo, hi = ci_wilson(x, n)`` still works.

def ci_wald(x: int, n: int, conf: float = 0.95):
    """Wald (normal-approximation) CI for a proportion. Poor near 0/1 and for small n."""
    p = x / n
    z = stats.norm.ppf(0.5 + conf / 2)
    half = z * math.sqrt(p * (1 - p) / n)
    return CI(lower=max(0.0, p - half), upper=min(1.0, p + half))


def ci_wilson(x: int, n: int, conf: float = 0.95):
    """Wilson score CI for a proportion -- the recommended general-purpose default."""
    p = x / n
    z = stats.norm.ppf(0.5 + conf / 2)
    z2 = z * z
    denom = 1 + z2 / n
    center = (p + z2 / (2 * n)) / denom
    half = (z / denom) * math.sqrt(p * (1 - p) / n + z2 / (4 * n * n))
    return CI(lower=max(0.0, center - half), upper=min(1.0, center + half))


def ci_clopper_pearson(x: int, n: int, conf: float = 0.95):
    """Exact (Clopper-Pearson) CI via Beta-distribution quantiles. Conservative but guaranteed coverage."""
    alpha = 1 - conf
    lower = 0.0 if x == 0 else stats.beta.ppf(alpha / 2, x, n - x + 1)
    upper = 1.0 if x == n else stats.beta.ppf(1 - alpha / 2, x + 1, n - x)
    return CI(lower=lower, upper=upper)


# --------------------------------------------------------------------------
# Confidence interval for a rate (exact Poisson)
# --------------------------------------------------------------------------
def ci_poisson_rate(events: int, person_time_total: float, conf: float = 0.95):
    """Exact Poisson (Garwood) CI for an incidence rate.

    Parameters
    ----------
    events : observed event count.
    person_time_total : denominator (sum of at-risk follow-up).
    conf : confidence level.

    Returns a CI named tuple with fields ``lower`` and ``upper`` -- bounds on
    the underlying rate ``lambda``. Unpacks like a tuple.
    """
    alpha = 1 - conf
    lo_count = 0.0 if events == 0 else stats.chi2.ppf(alpha / 2, 2 * events) / 2
    hi_count = stats.chi2.ppf(1 - alpha / 2, 2 * events + 2) / 2
    return CI(lower=lo_count / person_time_total,
              upper=hi_count / person_time_total)


def library_versions(x: int, n: int):
    from statsmodels.stats.proportion import proportion_confint

    return {
        "Wald (statsmodels)": proportion_confint(x, n, method="normal"),
        "Wilson (statsmodels)": proportion_confint(x, n, method="wilson"),
        "Clopper-Pearson (statsmodels)": proportion_confint(x, n, method="beta"),
        "Agresti-Coull (statsmodels)": proportion_confint(x, n, method="agresti_coull"),
    }


if __name__ == "__main__":
    print("=== Prevalence ===")
    print("120 cases in a population of 5000:", prevalence(120, 5000))

    print("\n=== Incidence ===")
    follow = [2.0, 5.0, 1.5, 5.0, 3.2, 5.0, 0.8, 4.1]  # person-years for 8 subjects
    pt = person_time(follow)
    print("person-time (years)    :", pt)
    print("incidence rate (3 events / py):", incidence_rate(3, pt), "per person-year")
    print("incidence proportion (3 / 8 at risk):", incidence_proportion(3, 8))

    print("\n=== 95% CIs for a proportion  x=8, n=100  (p = 0.08) ===")
    for name, fn in [("Wald", ci_wald), ("Wilson", ci_wilson), ("Clopper-Pearson", ci_clopper_pearson)]:
        lo, hi = fn(8, 100)
        print(f"{name:18s}: ({lo:.4f}, {hi:.4f})")

    print(f"\n=== 95% exact Poisson CI for a rate  3 events / {pt} person-years ===")
    lo, hi = ci_poisson_rate(3, pt)
    print(f"rate = {3 / pt:.4f}   CI = ({lo:.4f}, {hi:.4f}) per person-year")

    print("\n--- library (statsmodels.proportion_confint) ---")
    for k, v in library_versions(8, 100).items():
        print(f"{k:32s}: ({v[0]:.4f}, {v[1]:.4f})")
